src/main/xlat_func.c

```diff
@@ -1024,34 +1024,63 @@ static int fr_value_box_to_bin(TALLOC_CTX *ctx, REQUEST *request, uint8_t **out,
  *
  * Example: "%{md5:foo}" == "acbd18db4cc2f85cedef654fccc4a4d8"
  */
-static ssize_t md5_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			REQUEST *request, char const *fmt)
+static xlat_action_t xlat_md5(TALLOC_CTX *ctx, fr_cursor_t *out,
+			      REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+			      fr_value_box_t **in)
 {
 	uint8_t		digest[16];
-	size_t		i, len, inlen;
-	uint8_t		*p;
+	size_t		i, len;
 	FR_MD5_CTX	md5_ctx;
-	TALLOC_CTX	*tmp_ctx = NULL;
+	char		*buff;
+	fr_value_box_t	*vb_in;
+	fr_value_box_t	*vb_out;
 
-	VALUE_FROM_FMT(tmp_ctx, p, inlen, request, fmt);
+	/*
+	 * If there's no input, use empty value box so that md5 of null can be processed
+	 */
+	if (*in) {
+		if (fr_value_box_list_concat(ctx, *in, in, FR_TYPE_OCTETS, true) < 0) {
+			RPEDEBUG("Failed concatenating input");
+			return XLAT_ACTION_FAIL;
+		}
+		vb_in = *in;
+	} else {
+		MEM(vb_in = fr_value_box_alloc_null(ctx));
+	}
+
+	MEM(vb_out = fr_value_box_alloc_null(ctx));
 
 	fr_md5_init(&md5_ctx);
-	fr_md5_update(&md5_ctx, p, inlen);
+	fr_md5_update(&md5_ctx, vb_in->vb_octets, vb_in->vb_length);
```

Comment:
Probably simpler just to call fr_md5_update with NULL and 0, then you save the alloc.